mosaic-runtime/src/commonMain/kotlin/com/jakewharton/mosaic/mosaic.kt

```diff
@@ -155,21 +158,82 @@ internal class MosaicComposition(
 	coroutineScope: CoroutineScope,
 	private val terminalState: State<Terminal>,
 	private val keyEvents: ReceiveChannel<KeyEvent>,
-	onEndChanges: (MosaicNode) -> Unit,
+	private val onDrawing: (MosaicNode) -> Unit,
 ) {
 	private val job = Job(coroutineScope.coroutineContext[Job])
 	private val clock = BroadcastFrameClock()
 	private val composeContext: CoroutineContext = coroutineScope.coroutineContext + job + clock
 	val scope = CoroutineScope(composeContext)
 
-	private val applier = MosaicNodeApplier(onEndChanges)
+	private val applier = MosaicNodeApplier(::doLayout)
 	val rootNode = applier.root
 	private val recomposer = Recomposer(composeContext)
 	private val composition = Composition(applier, recomposer)
 
+	private val applyObserverHandle: ObserverHandle
+
+	private val readingStatesOnLayout = mutableSetOf<Any>()
+	private val readingStatesOnDrawing = mutableSetOf<Any>()
+
+	private val layoutBlockStateReadObserver: (Any) -> Unit = { readingStatesOnLayout.add(it) }
+	private val drawingBlockStateReadObserver: (Any) -> Unit = { readingStatesOnDrawing.add(it) }
+
+	@Volatile
+	private var needLayout = false
+
+	@Volatile
+	private var needDrawing = false
+
 	init {
 		GlobalSnapshotManager().ensureStarted(scope)
 		startRecomposer()
+		applyObserverHandle = registerSnapshotApplyObserver()
+		startListeningToNeedToLayoutOrDraw()
+	}
+
+	private fun doLayout(rootNode: MosaicNode) {
+		needLayout = false
+		Snapshot.observe(readObserver = layoutBlockStateReadObserver) {
+			rootNode.measureAndPlace()
+		}
+		doDrawing(rootNode)
+	}
+
+	private fun doDrawing(rootNode: MosaicNode) {
```

Comment:
Here and everywhere else, I would prefer "draw" and not "drawing" just to agree more with "layout" as a noun.

Also maybe "perform" instead of "do". There's nothing technically wrong with "do", it's just kind of weak as far as verbs go.